misc-utils/lsfd-sock-xinfo.c

```diff
@@ -52,41 +55,132 @@ static void load_xinfo_from_proc_udp6(ino_t netns_inode);
 static void load_xinfo_from_proc_udplite6(ino_t netns_inode);
 static void load_xinfo_from_proc_raw6(ino_t netns_inode);
 static void load_xinfo_from_proc_netlink(ino_t netns_inode);
+static void load_xinfo_from_proc_packet(ino_t netns_inode);
 
 static int self_netns_fd = -1;
 static struct stat self_netns_sb;
 
 static void *xinfo_tree;	/* for tsearch/tfind */
 static void *netns_tree;
 
+struct iface {
+	unsigned int index;
+	const char *name;
+};
+
+static const char *get_iface_name(ino_t netns, unsigned int iface_index);
+
+static int iface_compare(const void *a, const void *b)
+{
+       const struct iface *iface_a = a;
+       const struct iface *iface_b = b;
+
+       if (iface_a->index < iface_b->index)
+	       return -1;
+       else if (iface_a->index > iface_b->index)
+	       return 1;
+       else
+	       return 0;
+}
+
+static void iface_free(void * iface)
+{
+       free((void *)(((struct iface *)iface)->name));
+       free(iface);
+}
+
+struct netns {
+	ino_t inode;
+	void* iface_tree;
+};
+
 static int netns_compare(const void *a, const void *b)
 {
-	if (*(ino_t *)a < *(ino_t *)b)
+	const struct netns *netns_a = a;
+	const struct netns *netns_b = b;
+
+	if (netns_a->inode < netns_b->inode)
 		return -1;
-	else if (*(ino_t *)a > *(ino_t *)b)
+	else if (netns_a->inode > netns_b->inode)
 		return 1;
 	else
 		return 0;
 }
 
+static void netns_free(void *netns)
+{
+	struct netns *nsobj = netns;
+
+	tdestroy(nsobj->iface_tree, iface_free);
+	free(netns);
+}
+
+/*
+ * iface index -> iface name mappings
+ */
+static void load_ifaces_from_getifaddrs(struct netns *nsobj)
+{
+	struct ifaddrs *ifa_list;
+
+	if (getifaddrs(&ifa_list) < 0)
+		return;
+
+	for (struct ifaddrs *ifa = ifa_list; ifa != NULL; ifa = ifa->ifa_next) {
+		struct iface *iface;
+		unsigned int if_index = if_nametoindex(ifa->ifa_name);
```

Comment:
Instead of having all of this code managing a tree, wouldn't it be easier to just have an array?

Maybe even faster because of less pointers, the name could also be a data member of `struct iface` instead of a pointer.

Reply:
> Instead of having all of this code managing a tree, wouldn't it be easier to just have an array?

Network interfaces are stored in a tree in a network namespace.
Do you mean it is easier that the network interfaces are stored in an array in a network namespace?

Reply:
Yes, but this is subjective.
Something like:

```c
for (ifa = ifa_list; ifa; ifa = ifa->next)
    count++;

ret = xcalloc(count + 1, sizeof(*iface));

struct iface *iface = ret;
for (ifa = ifa_list; ifa; ifa = ifa->next) {
    strncpy(iface, ifa->name, sizeof(iface->name));
    ifa->index = if_nametoindex(ifa->name);
    iface++;
}
```

Reply:
I understand that your code is better.
I could not find the technique using ifa_list twice, one for counting and one for copying.
I thought of using a list instead of an array.
I will update my change.